gcf_data_mapper/parsers/family.py

```diff
@@ -3,6 +3,132 @@
+    result_areas = row.at[FamilyColumnsNames.RESULT_AREAS.value]
+
+    area_key = FamilyNestedColumnNames.AREA.value
+    name_key = FamilyNestedColumnNames.NAME.value
+    region_key = FamilyNestedColumnNames.REGION.value
+    type_key = FamilyNestedColumnNames.TYPE.value
+
+    co_financing_budgets = get_budgets(
+        funding_sources, GCFProjectBudgetSource.CO_FINANCING.value
+    )
+    gcf_budgets = get_budgets(funding_sources, GCFProjectBudgetSource.GCF.value)
+
+    implementing_agencies = [entity[name_key] for entity in entities]
+    regions = [country[region_key] for country in countries]
+    areas = [result[area_key] for result in result_areas]
+    types = [result[type_key] for result in result_areas]
+
+    # As we are filtering the budget information by source for gcf and co financing, we
+    # know there will be instances where only one type of funding exists so checking
+    # for empty/false values would create false positives, hence the exclusion from this
+    # check
+    if arrays_contain_empty_values(
+        [
+            ("Implementing Agencies", implementing_agencies),
+            ("Regions", regions),
+            ("Result Areas", areas),
+            ("Result Types", types),
+        ],
+        row.at[FamilyColumnsNames.PROJECTS_ID.value],
+    ):
+        return None
+
+    metadata = {
+        "approved_ref": [row.at[FamilyColumnsNames.APPROVED_REF.value]],
+        "implementing_agencies": list(set(implementing_agencies)),
+        "project_id": [row.at[FamilyColumnsNames.PROJECTS_ID.value]],
+        "project_url": [row.at[FamilyColumnsNames.PROJECT_URL.value]],
+        "project_value_fund_spend": gcf_budgets,
+        "project_value_co_financing": co_financing_budgets,
+        "regions": list(set(regions)),
+        "result_areas": list(set(areas)),
+        "result_types": list(set(types)),
+        "sector": [row.at[FamilyColumnsNames.SECTOR.value]],
+        "theme": [row.at[FamilyColumnsNames.THEME.value]],
+    }
+
+    return metadata
+
+
+def process_row(row: pd.Series) -> Optional[dict]:
+    """Map the family data based on the provided row.
+
+    :param pd.Series row: The row containing family information.
+    :return Optional[dict]: A dictionary containing mapped data for the family entity.
+        The function will return None, if the row contains missing data from expected columns/fields
+    """
+
+    required_columns = [column.value for column in FamilyColumnsNames]
+
+    doc_id = (
+        row.at[FamilyColumnsNames.PROJECTS_ID.value]
+        if FamilyColumnsNames.PROJECTS_ID.value in row.index
+        and pd.notna(row.at[FamilyColumnsNames.PROJECTS_ID.value])
+        else None
+    )
+
+    if not doc_id:
+        click.echo("🛑 Skipping row as it does not contain a project id")
+        return None
+
+    # Check for missing required columns in the row and raise error accordingly
+    check_row_for_missing_columns(row, required_columns, doc_id)
```

Comment:
We need to be checking for missing columns in the dataframe a level up from here in the main `family` function - as if we do the check here it will raise an attribute error and crash the whole program. One row will not be missing the column (not possible, as the dataframe would have failed to build as there would be a column mismatch) -> however one row may contain NA values for a column. And we're already checking for this below :)